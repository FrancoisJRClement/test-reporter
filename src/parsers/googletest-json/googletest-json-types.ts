/// reflects documentation at https://github.com/google/googletest/blob/main/docs/advanced.md

export interface GoogleTestFailure {
  failure: string
  type: string | 'Unknown type'
}

export interface GoogleTestCase {
  name: string
  file: string
  line: number
  status: 'RUN' | 'NOTRUN'
  result?: 'COMPLETED' | string
  timestamp?: string
  time: string
  classname: string
  failures: GoogleTestFailure[]
}

export interface GoogleTestGroup {
  name: string
  tests: number
  failures:number
  disabled: number
  errors?: number
  timestamp?: string
  time: string
  testsuite: GoogleTestCase[]
}

export interface GoogleTestRun {
  name: string
  tests: number
  failures:number
  disabled: number
  errors: number
  timestamp?: string
  time: string
  testsuites: GoogleTestGroup[]
}