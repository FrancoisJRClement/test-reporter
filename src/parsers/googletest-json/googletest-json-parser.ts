import {ParseOptions, TestParser} from '../../test-parser'

import {getBasePath, normalizeFilePath} from '../../utils/path-utils'

import {GoogleTestCase, GoogleTestGroup, GoogleTestRun} from './googletest-json-types'

import {
  TestExecutionResult,
  TestRunResult,
  TestSuiteResult,
  TestGroupResult,
  TestCaseResult,
  TestCaseError
} from '../../test-results'

export class GoogleTestJsonParser implements TestParser {
  assumedWorkDir: string | undefined

  constructor(readonly options: ParseOptions) {}

  async parse(path: string, content: string): Promise<TestRunResult> {
    try {
      const testRun = JSON.parse(content) as GoogleTestRun
      const runResult = this.getTestRunResult(path, testRun)
      return Promise.resolve(runResult)
    } catch (parseException) {
      if (parseException instanceof SyntaxError)
        throw new Error('Error while parsing file "' + path + '":\n\t' + parseException.message)
      else throw parseException
    }
  }

  private getTestRunResult(path: string, testRun: GoogleTestRun): TestRunResult {
    const totalTime = this.getTime(testRun.time)

    const suites = [new TestSuiteResult(testRun.name, this.getGroups(testRun.testsuites), totalTime)]

    return new TestRunResult(this.getRelativePath(path), suites, totalTime)
  }

  private getGroups(testGroups: GoogleTestGroup[]): TestGroupResult[] {
    return testGroups.map(testGroup => new TestGroupResult(testGroup.name, this.getCases(testGroup.testsuite)))
  }

  private getCases(testCases: GoogleTestCase[]): TestCaseResult[] {
    return testCases.map(testCase => {
      const result = this.getResult(testCase)
      if (result === 'failed') {
        return new TestCaseResult(testCase.name, result, this.getTime(testCase.time), this.getError(testCase))
      } else {
        return new TestCaseResult(testCase.name, result, this.getTime(testCase.time))
      }
    })
  }

  private getTime(time: string): number {
    const fullLength = time.length
    const valueLength = fullLength - 1
    if (time[valueLength] !== 's') {
      return NaN
    } else {
      return Number(time.slice(0, valueLength)) * 1000
    }
  }

  private getResult(testCase: GoogleTestCase): TestExecutionResult {
    if (testCase.status !== 'RUN') {
      return 'skipped'
    } else if (testCase.failures === undefined || testCase.failures.length === 0) {
      return 'success'
    } else {
      return 'failed'
    }
  }

  private getError(testCase: GoogleTestCase): TestCaseError | undefined {
    if (testCase.failures.length === 0) {
      return undefined
    } else {
      const errorMessage = testCase.failures.map(testFailure => testFailure.failure).join('\n\n')
      return {details: errorMessage}
    }
  }

  private getRelativePath(path: string): string {
    const prefix = 'file://'
    if (path.startsWith(prefix)) {
      path = path.substr(prefix.length)
    }

    path = normalizeFilePath(path)
    const workDir = this.getWorkDir(path)
    if (workDir !== undefined && path.startsWith(workDir)) {
      path = path.substr(workDir.length)
    }
    return path
  }

  private getWorkDir(path: string): string | undefined {
    return (
      this.options.workDir ??
      this.assumedWorkDir ??
      (this.assumedWorkDir = getBasePath(path, this.options.trackedFiles))
    )
  }
}
