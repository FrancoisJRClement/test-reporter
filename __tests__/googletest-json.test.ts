import * as fs from 'fs'
import * as path from 'path'

import {GoogleTestJsonParser} from '../src/parsers/googletest-json/googletest-json-parser'
import {ParseOptions} from '../src/test-parser'
import {getReport} from '../src/report/get-report'
import {normalizeFilePath} from '../src/utils/path-utils'

describe('googletest-json tests', () => {
  it('produces empty empty test run when there is no fixture instanciation ', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'empty', 'googletest-json.json')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const trackedFiles = ['testClogRecord.cpp']
    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles
    }

    const parser = new GoogleTestJsonParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result.tests).toBe(0)
    expect(result.result).toBe('success')
  })

  it('report from googletest test results matches snapshot', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'googletest-json.json')
    const outputPath = path.join(__dirname, '__outputs__', 'googletest-json.md')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const trackedFiles = ['testClogRecord.cpp']
    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles
    }

    const parser = new GoogleTestJsonParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result).toMatchSnapshot()

    const report = getReport([result])
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, report)
  })
})
