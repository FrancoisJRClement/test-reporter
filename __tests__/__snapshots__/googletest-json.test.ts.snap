// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`googletest-json tests report from googletest test results matches snapshot 1`] = `
TestRunResult {
  "path": "fixtures/googletest-json.json",
  "suites": [
    TestSuiteResult {
      "groups": [
        TestGroupResult {
          "name": "testClogRecord_simple",
          "tests": [
            TestCaseResult {
              "error": undefined,
              "name": "default_constructors",
              "result": "success",
              "time": 0,
            },
            TestCaseResult {
              "error": undefined,
              "name": "fatal",
              "result": "success",
              "time": 0,
            },
            TestCaseResult {
              "error": undefined,
              "name": "critical",
              "result": "success",
              "time": 0,
            },
            TestCaseResult {
              "error": undefined,
              "name": "warning",
              "result": "success",
              "time": 0,
            },
            TestCaseResult {
              "error": undefined,
              "name": "notice",
              "result": "success",
              "time": 0,
            },
            TestCaseResult {
              "error": undefined,
              "name": "progress",
              "result": "success",
              "time": 0,
            },
            TestCaseResult {
              "error": undefined,
              "name": "debug",
              "result": "success",
              "time": 0,
            },
            TestCaseResult {
              "error": undefined,
              "name": "extension",
              "result": "success",
              "time": 0,
            },
            TestCaseResult {
              "error": undefined,
              "name": "DISABLED_some_disabled",
              "result": "skipped",
              "time": 0,
            },
            TestCaseResult {
              "error": {
                "details": "../../../test/testClogRecord.cpp:342
Expected equality of these values:
  379
  dSomeValue
    Which is: 2.235


../../../test/testClogRecord.cpp:343
Expected equality of these values:
  "Expecting this..."
  sSomeText
    Which is: "Got that!"
",
              },
              "name": "some_failures",
              "result": "failed",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "testClogRecord_workingDir",
          "tests": [
            TestCaseResult {
              "error": undefined,
              "name": "record_streaming",
              "result": "success",
              "time": 1,
            },
            TestCaseResult {
              "error": undefined,
              "name": "extension_streaming",
              "result": "success",
              "time": 12,
            },
            TestCaseResult {
              "error": {
                "details": "../../../test/testClogRecord.cpp:356
Expected equality of these values:
  379
  dSomeValue
    Which is: 2.235


../../../test/testClogRecord.cpp:357
Expected equality of these values:
  "Expecting this..."
  sSomeText
    Which is: "Got that!"
",
              },
              "name": "other_failures",
              "result": "failed",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "DISABLED_testClogRecord_simple",
          "tests": [
            TestCaseResult {
              "error": undefined,
              "name": "skipped_failure",
              "result": "skipped",
              "time": 0,
            },
            TestCaseResult {
              "error": undefined,
              "name": "skipped_success",
              "result": "skipped",
              "time": 0,
            },
          ],
        },
      ],
      "name": "AllTests",
      "totalTime": 14,
    },
  ],
  "totalTime": 14,
}
`;
